caffe2/python/optimizer.py

```diff
@@ -168,6 +167,61 @@ def scale_learning_rate(self, scale):
         self.base_learning_rate *= scale
         return
 
+class MultiPrecisionSgdOptimizer(SgdOptimizer):
+    def __init__(self, base_learning_rate, momentum, policy, **kwargs):
+        super(SgdOptimizer, self).__init__() # self, base_learning_rate, **kwargs)
+        self.base_learning_rate = base_learning_rate
+        self.momentum = momentum
+        self.policy = policy
+        self.init_kwargs = kwargs
+
+    def _run(self, net, param_init_net, param_info):
+        param = param_info.blob
+        param_fp32 = param_info.blob_copy['float'] if param_info.blob_copy is not None else None
+
+        # If we have a straight fp32 parameter, run the base class
+        if param_fp32 == None:
+            return SgdOptimizer._run(self, net, param_init_net, param_info)
+
+        grad = param_info.grad
+        if self.base_learning_rate <= 0:
+            return
+
+        lr, _ = self.build_lr(
+            net, param_init_net,
+            base_learning_rate=-self.base_learning_rate,
+            learning_rate_blob=str(param) + "_lr",
+            policy=self.policy,
+            **(self.init_kwargs)
+        )
+
+        ONE = param_init_net.ConstantFill([], "ONE", shape=[1], value=1.0)
+        self._aux_params.shared.append(ONE)
+
+        momentum_data = param_init_net.ConstantFill(
+            [param], param + "_momentum", value=0.)
+        self._aux_params.local.append(momentum_data)
+
+        assert not isinstance(grad, core.GradientSlice), \
```

Comment:
I don't really see how apart from abstracting out some more of the details (which I consider out of the scope of this patch set) -- I already call out to `SgdOptimizer` in the case of fp32 parameters.